From 3ed524feda787229cef8076048e00c7933218ced Mon Sep 17 00:00:00 2001
From: d34d <clark@cyngn.com>
Date: Wed, 4 Jan 2017 10:29:34 -0800
Subject: [PATCH 02/10] Introduce sepolicy exceptions for theme assets

Assets such as composed icons and ringtones need to be accessed
by apps. This patch adds the policy needed to facilitate this.

Change-Id: I0420de579aed0cff5add181cd0a8bf0f2b05d723
---
 app.te         | 4 ++++
 bootanim.te    | 4 ++++
 file.te        | 3 +++
 file_contexts  | 3 +++
 mediaserver.te | 4 ++++
 system_app.te  | 4 ++++
 zygote.te      | 4 ++++
 7 files changed, 26 insertions(+)

diff --git a/app.te b/app.te
index 19a7dac..3a2878d 100644
--- a/app.te
+++ b/app.te
@@ -468,3 +468,7 @@ neverallow {
 # Foreign dex profiles are just markers. Prevent apps to do anything but touch them.
 neverallow appdomain user_profile_foreign_dex_data_file:file rw_file_perms;
 neverallow appdomain user_profile_foreign_dex_data_file:dir { open getattr read ioctl remove_name };
+
+# Themed resources (i.e. composed icons)
+allow appdomain theme_data_file:dir r_dir_perms;
+allow appdomain theme_data_file:file r_file_perms;
diff --git a/bootanim.te b/bootanim.te
index c3091ab..3ae9478 100644
--- a/bootanim.te
+++ b/bootanim.te
@@ -32,3 +32,7 @@ r_dir_file(bootanim, cgroup)
 
 # System file accesses.
 allow bootanim system_file:dir r_dir_perms;
+
+# Themed resources (bootanimation)
+allow bootanim theme_data_file:dir search;
+allow bootanim theme_data_file:file r_file_perms;
diff --git a/file.te b/file.te
index dfa3c9b..1e18b89 100644
--- a/file.te
+++ b/file.te
@@ -266,3 +266,6 @@ allow postinstall_file self:filesystem associate;
 # Should be:
 #   type apk_data_file, file_type, data_file_type;
 neverallow fs_type file_type:filesystem associate;
+
+# Themes
+type theme_data_file, file_type, data_file_type;
diff --git a/file_contexts b/file_contexts
index 3448f21..392e655 100644
--- a/file_contexts
+++ b/file_contexts
@@ -401,3 +401,6 @@
 /mnt/user(/.*)?             u:object_r:mnt_user_file:s0
 /mnt/runtime(/.*)?          u:object_r:storage_file:s0
 /storage(/.*)?              u:object_r:storage_file:s0
+
+# Themes
+/data/system/theme(/.*)?  u:object_r:theme_data_file:s0
diff --git a/mediaserver.te b/mediaserver.te
index dc05e14..661bdee 100644
--- a/mediaserver.te
+++ b/mediaserver.te
@@ -139,3 +139,7 @@ neverallow mediaserver { file_type fs_type }:file execute_no_trans;
 
 # do not allow privileged socket ioctl commands
 neverallowxperm mediaserver domain:{ rawip_socket tcp_socket udp_socket } ioctl priv_sock_ioctls;
+
+# Themed resources (i.e. composed icons)
+allow mediaserver theme_data_file:dir r_dir_perms;
+allow mediaserver theme_data_file:file r_file_perms;
diff --git a/system_app.te b/system_app.te
index 50320c5..afd65d7 100644
--- a/system_app.te
+++ b/system_app.te
@@ -75,3 +75,7 @@ allow system_app sysfs_zram:dir search;
 allow system_app sysfs_zram:file r_file_perms;
 
 control_logd(system_app)
+
+# Themes
+allow system_app theme_data_file:dir create_dir_perms;
+allow system_app theme_data_file:file create_file_perms;
diff --git a/zygote.te b/zygote.te
index c6b343c..c650c17 100644
--- a/zygote.te
+++ b/zygote.te
@@ -104,3 +104,7 @@ neverallow zygote {
   data_file_type
   -dalvikcache_data_file # map PROT_EXEC
 }:file no_x_file_perms;
+
+# Themes
+allow zygote theme_data_file:file r_file_perms;
+allow zygote theme_data_file:dir r_dir_perms;
-- 
2.11.1

